### PR TITLE
Mild refactor of the tooltip system

### LINKED ISSUE #2284: Rename chartLegendPhysConds to chartLegendPhysicalConditions
and remove the warning suppression. I didn't do this in my spelling check enabling PR because these seem to be used from a Godot scene, so this needs a little bit extra care to correctly rename it.

### PR DESCRIPTION
**Pull request by Kasterisk**
**Brief Description of What This PR Does**

This refactors the tooltip system in order to fix some quirky behaviors functionality-wise and code-wise that I found while browsing through tooltip related codes (and some playtesting). List of changes in addition to the ones covered by the commit messages:
- Added signal Disconnect calls on the tooltip signals, this is needed to completely unregister tooltip from a control (due to the actual callback data instance still lingering on for a bit). Forgot to do this in previous tooltip PR after changing `ToolTipCallbackData`'s base class back to Reference.
- Fixed tooltip display timer not resetting on mouse button press when tooltip is not visible.
- Consolidate the tooltip visibility transitioning functionality to make it less error prone and more maintainable, still need further improvements.
- Tweaked the display delay of some editor tooltips.
- Minor general UI and code tweaks.

**Related Issues**

Closes #2284 

**Progress Checklist**

Note: before starting this checklist the issue should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [ ] Initial code review passed
- [ ] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author.

### CODE REVIEW

File: src/gui_common/tooltip/ToolTipManager.cs [hunk 1]
```diff
@@ -56,7 +67,15 @@ public bool Display
...
-            UpdateToolTipVisibility();
+
+            if (previousToolTip != null)
+                FinalizeToolTipVisibility(previousToolTip, false);
+
+            if (display)
+            {
+                // Set timer
+                displayTimer = MainToolTip.DisplayDelay;
+            }
```
Comment (FuJa0815) [inline src/gui_common/tooltip/ToolTipManager.cs:78]: is it beneficial to run this code when display was already true?

File: src/gui_common/tooltip/ToolTipManager.cs [hunk 2]
```diff
@@ -296,33 +321,56 @@ private void FetchToolTips()
...
-    private void UpdateToolTipVisibility()
+    private void FinalizeToolTipVisibility(ICustomToolTip tooltip, bool visible)
     {
-        // TODO: Fix the current tooltip changing while still fading out
-        // when quickly mousing over multiple closely positioned elements
-        // (Happens when a single tooltip is registered to multiple Controls)
-
-        // Make sure to hide any other tooltips that are still visible
-        HideAllToolTips();
+        if ((tooltip.ToolTipNode.Visible && visible) || (!tooltip.ToolTipNode.Visible && !visible))
```
Comment (FuJa0815) [inline src/gui_common/tooltip/ToolTipManager.cs:334]:
```suggestion
        if (tooltip.ToolTipNode.Visible == visible)
```

File: src/gui_common/tooltip/ToolTipManager.cs [hunk 3]
```diff
@@ -296,33 +321,56 @@ private void FetchToolTips()
...
+                {
+                    GUICommon.Instance.ModulateFadeIn(tooltip.ToolTipNode, Constants.TOOLTIP_FADE_SPEED);
+                }
+                else
+                {
+                    GUICommon.Instance.ModulateFadeOut(tooltip.ToolTipNode, Constants.TOOLTIP_FADE_SPEED);
+                }
+
+                break;
+            }
```
Comment (FuJa0815) [inline src/gui_common/tooltip/ToolTipManager.cs:359]: use `case ToolTipTransitioning.Fade when visible:` and `case ToolTipTransitioning.Fade when !visible:` here

File: src/gui_common/tooltip/ToolTipManager.cs (hunk 1 above)
Comment (Kasterisk) [reply src/gui_common/tooltip/ToolTipManager.cs:78]: Yes, consider one of a likely scenario: `MainToolTip` was reassigned to another tooltip from an initial tooltip with `Display` already set true, and assuming the newly assigned tooltip is hidden we could make it visible by setting `Display` to true once more (basically to reset the display timer).

File: src/gui_common/tooltip/ToolTipManager.cs [hunk 4]
```diff
@@ -56,7 +67,15 @@ public bool Display
...
             display = value;
-            UpdateToolTipVisibility();
+
+            if (previousToolTip != null)
+                FinalizeToolTipVisibility(previousToolTip, false);
+
+            if (display)
+            {
+                // Set timer
+                displayTimer = MainToolTip.DisplayDelay;
```
Comment (FuJa0815) [inline src/gui_common/tooltip/ToolTipManager.cs:77]: If I read this correctly setting `Display` to true doesn't mean that the tooltip does get displayed but that the tooltip will be displayed in some time, right? If that's the case, I don't think that `Display` is a very descriptive name anymore

File: src/microbe_stage/editor/MicrobeEditor.tscn [hunk 5]
```diff
@@ -1435,45 +1437,39 @@ SelectionGroup = SubResource( 19 )
 [node name="bindingAgent" parent="MicrobeEditorGUI/CellEditor/LeftPanel/MainPanel/VBoxContainer/MarginContainer/EditingPanel/Structure/ScrollContainer/MarginContainer/PartsSelection/Organelle/Clip/GridContainer" index="1" groups=[
 "PlaceablePartSelectionElement",
 ] instance=ExtResource( 12 )]
-margin_right = 98.0
```
Comment (FuJa0815) [inline src/microbe_stage/editor/MicrobeEditor.tscn:1438]: Why did you change the organelle order in the editor? It doesn't change anything ingame, but is this an unintened change?

File: src/gui_common/tooltip/ToolTipManager.cs (hunk 4 above)
Comment (Kasterisk) [reply src/gui_common/tooltip/ToolTipManager.cs:77]: Well, it still _displays_ the tooltip, if the tooltip has its delay set to zero it will get displayed right away. I think this can be mentioned in the summary that the display time depends on set delay as there's no mention of this currently.

File: src/microbe_stage/editor/MicrobeEditor.tscn (hunk 5 above)
Comment (Kasterisk) [reply src/microbe_stage/editor/MicrobeEditor.tscn:1438]: Automatic change, will try to see if these are revertable.

Comment (Kasterisk) [reply src/microbe_stage/editor/MicrobeEditor.tscn:1438]: Tried reverting it and seems not possible, Godot is persistent with the changes the second I modified the editor tabs visibility.

Comment (hhyyrylainen) [reply src/microbe_stage/editor/MicrobeEditor.tscn:1438]: Updating the margins, happens pretty often. It seems that very often Godot doesn't realize the changes it wants to make automatically *unless* you restart Godot.

File: src/gui_common/tooltip/ToolTipManager.cs [hunk 6]
```diff
@@ -58,8 +58,8 @@ public ICustomToolTip MainToolTip
     }
 
     /// <summary>
-    ///   Displays the current tooltip if set true. It's preferable to set this
-    ///   rather than directly from the tooltip
+    ///   If true displays the current tooltip with a set delay <see cref="ICustomToolTip.DisplayDelay"/>.
```
Comment (FuJa0815) [inline src/gui_common/tooltip/ToolTipManager.cs:61]:
```suggestion
    ///   If true displays the current tooltip with a set delay of <see cref="ICustomToolTip.DisplayDelay"/>.
```

File: src/gui_common/tooltip/ICustomToolTip.cs [hunk 7]
```diff
@@ -16,57 +16,56 @@ public enum ToolTipPositioning
...
+{
+    /// <summary>
+    ///   Immediately display and hide the tooltip without animation.
+    /// </summary>
+    Immediate,
+
+    /// <summary>
+    ///   Use fading to display and hide the tooltip.
+    /// </summary>
+    Fade,
```
Comment (FuJa0815) [inline src/gui_common/tooltip/ICustomToolTip.cs:32]: Actually, while playtesting I haven't found a tooltip which fades either in this PR or in the master. Did I miss some? Aren't there any? Is this broken?

File: src/gui_common/tooltip/ToolTipManager.cs [hunk 8]
```diff
@@ -1,9 +1,9 @@
-﻿using System;
+using System;
```
Comment (FuJa0815) [inline src/gui_common/tooltip/ToolTipManager.cs:1]: @hhyyrylainen this file has lost the BOM but CI is not complaining.

Comment (hhyyrylainen) [reply src/gui_common/tooltip/ToolTipManager.cs:1]: I noticed that, and a fix for the CI not actually running on the right files is on my branch (it's waiting for the feature to be done before merging): https://github.com/Revolutionary-Games/Thrive/pull/2439

Comment (FuJa0815) [reply src/gui_common/tooltip/ToolTipManager.cs:1]: @Kasterisk
you can add the BOM to the file like [that on unix](https://stackoverflow.com/a/35807131) or like [that on windows with np++](https://stackoverflow.com/a/8432691)

File: src/gui_common/tooltip/ICustomToolTip.cs (hunk 7 above)
Comment (Kasterisk) [reply src/gui_common/tooltip/ICustomToolTip.cs:32]: There currently aren't any.

File: src/gui_common/tooltip/ICustomToolTip.cs [hunk 9]
```diff
@@ -16,57 +16,56 @@ public enum ToolTipPositioning
     FollowMousePosition,
 }
 
+/// <summary>
+///   How a tooltip should transition on display and hide.
```
Comment (hhyyrylainen) [inline src/gui_common/tooltip/ICustomToolTip.cs:20]:
```suggestion
///   How a tooltip should transition on becoming visible and on being hidden.
```

File: src/gui_common/tooltip/ToolTipManager.cs [hunk 10]
```diff
@@ -296,33 +321,48 @@ private void FetchToolTips()
...
-            // Set timer
-            displayTimer = MainToolTip.DisplayDelay;
+            case ToolTipTransitioning.Immediate:
+                tooltip.ToolTipNode.Visible = visible;
+                break;
+            case ToolTipTransitioning.Fade when visible:
+                GUICommon.Instance.ModulateFadeIn(tooltip.ToolTipNode, Constants.TOOLTIP_FADE_SPEED);
+                break;
+
+            // ReSharper disable once ConditionIsAlwaysTrueOrFalse
```
Comment (hhyyrylainen) [inline src/gui_common/tooltip/ToolTipManager.cs:349]: Why's this here?
I'm about to open a PR where the jetbrains version is updated to 2021.2, might that help?

Comment (hhyyrylainen) [reply src/gui_common/tooltip/ToolTipManager.cs:349]: Also first time when I see this `when` syntax. So if the check doesn't pass otherwise how about combining these to a one case with an if inside it?

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]

File: src/gui_common/tooltip/ToolTipManager.cs (hunk 10 above)
Comment (Kasterisk) [reply src/gui_common/tooltip/ToolTipManager.cs:349]: Just knew about it as well but CI of course just doesn't like it. The if block inside one case is the original form I intended, will just revert to that.

**Comment by sirdanforth** [comment]
I did not find any bugs or errors pertaining to this PR.

File: src/gui_common/tooltip/ToolTipManager.cs (hunk 10 above)
Comment (FuJa0815) [reply src/gui_common/tooltip/ToolTipManager.cs]: Pretty sure the jb complaint was significant. removing the `when !visible` probably should have fixed it. This is because the case above already handles the fade case when visible, so here the when condition is unnecessary.

**Review by hhyyrylainen — APPROVED** [review]
